### ISSUE TITLE
Yolov8CPPInference fails to load ONNX model, throws reshape error despite opset=12 and imgsz=640 set

### ISSUE THREAD

**Issue by sonali20** [ultralytics/ultralytics#6031]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Predict

### Bug

I'm simply running the example code in [YOLOv8-CPP-Inference](https://github.com/ultralytics/ultralytics/tree/main/examples/YOLOv8-CPP-Inference) on my machine. I modified the CMakelists.txt to comment out the CUDA section, set `runOnGpu` to false in `main.cpp` and modified the `projectBasePath` to point to my ultralytics folder. These are the only changes I made to the code. 

I'm using OpenCV 4.7.0 (since it failed to work with earlier versions). 

I downloaded the yolo model with the command -
`yolo export model=yolov8s.pt format=onnx opset=12 imgsz=640`

and got -
```
Downloading https://github.com/ultralytics/assets/releases/download/v0.0.0/yolov8s.pt to 'yolov8s.pt'...
100%|██████████████████████████████████████████████████████████████████████████████████████████████████| 21.5M/21.5M [00:02<00:00, 11.0MB/s]
Ultralytics YOLOv8.0.188 🚀 Python-3.8.10 torch-2.0.1+cpu CPU (AMD Ryzen 7 PRO 5850U with Radeon Graphics)
YOLOv8s summary (fused): 168 layers, 11156544 parameters, 0 gradients

PyTorch: starting from 'yolov8s.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 84, 8400) (21.5 MB)

ONNX: starting export with onnx 1.14.1 opset 12...
============== Diagnostic Run torch.onnx.export version 2.0.1+cpu ==============
verbose: False, log level: Level.ERROR
======================= 0 NONE 0 NOTE 0 WARNING 0 ERROR ========================

ONNX: export success ✅ 0.6s, saved as 'yolov8s.onnx' (42.8 MB)

Export complete (2.2s)
Results saved to /home/sonali/savioke/ultralytics
Predict:         yolo predict task=detect model=yolov8s.onnx imgsz=640  
Validate:        yolo val task=detect model=yolov8s.onnx imgsz=640 data=coco.yaml  
Visualize:       https://netron.app
💡 Learn more at https://docs.ultralytics.com/modes/export
```

But when I run the binary I get this error -
```
Running on CPU
[ERROR:0@0.070] global net_impl.cpp:1161 getLayerShapesRecursively OPENCV/DNN: [Reshape]:(onnx_node!/model.22/dfl/Reshape): getMemoryShapes() throws exception. inputs=1 outputs=1/1 blobs=0
[ERROR:0@0.070] global net_impl.cpp:1167 getLayerShapesRecursively     input[0] = [ 1 64 6300 ]
[ERROR:0@0.070] global net_impl.cpp:1171 getLayerShapesRecursively     output[0] = [ ]
[ERROR:0@0.070] global net_impl.cpp:1177 getLayerShapesRecursively Exception message: OpenCV(4.7.0) /home/sonali/savioke/opencv/modules/dnn/src/layers/reshape_layer.cpp:109: error: (-215:Assertion failed) total(srcShape, srcRange.start, srcRange.end) == maskTotal in function 'computeShapeByReshapeMask'

terminate called after throwing an instance of 'cv::Exception'
  what():  OpenCV(4.7.0) /home/sonali/savioke/opencv/modules/dnn/src/layers/reshape_layer.cpp:109: error: (-215:Assertion failed) total(srcShape, srcRange.start, srcRange.end) == maskTotal in function 'computeShapeByReshapeMask'

Aborted (core dumped)
```
Previous tickets that document this error seemed to have been solved by specifying the opset but it isn't working in this case

### Environment

```
Ultralytics YOLOv8.0.188 🚀 Python-3.8.10 torch-2.0.1+cpu CPU (AMD Ryzen 7 PRO 5850U with Radeon Graphics)
Setup complete ✅ (16 CPUs, 30.2 GB RAM, 28.3/144.3 GB disk)

OS                  Linux-5.15.0-87-generic-x86_64-with-glibc2.29
Environment         Linux
Python              3.8.10
Install             pip
RAM                 30.18 GB
CPU                 AMD Ryzen 7 PRO 5850U with Radeon Graphics
CUDA                None

matplotlib          ✅ 3.7.3>=3.3.0
numpy               ✅ 1.23.3>=1.22.2
opencv-python       ✅ 4.8.0.76>=4.6.0
pillow              ✅ 10.0.1>=7.1.2
pyyaml              ✅ 5.3.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.10.1>=1.4.1
torch               ✅ 2.0.1+cpu>=1.8.0
torchvision         ✅ 0.15.2+cpu>=0.9.0
tqdm                ✅ 4.66.1>=4.64.0
pandas              ✅ 2.0.3>=1.1.4
seaborn             ✅ 0.12.2>=0.11.0
psutil              ✅ 5.5.1
py-cpuinfo          ✅ 9.0.0
```

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
Thank you for your detailed issue report, @sonali20. The error implies there's a problem when trying to reshape the input tensor in the model after loading the ONNX model with OpenCV's DNN module. Your ONNX model is expecting a different size from what it's receiving.

As for the issue, first double-check that you are using the correct image size when you call the prediction. The dimensions should match the dimensions when exporting the model, which in your case is 640 (as per your given command `yolo export model=yolov8s.pt format=onnx opset=12 imgsz=640`).

Second, even though you have set the opset and the imgsz, the error could be related to the OpenCV version you're using. You mentioned that you're using OpenCV 4.7.0, but as of now, we don't have confirmation about support or issues for that specific version.

Finally, if the problem persists, try to simplify the possible issues by running additional tests. Use different models, opsets and input sizes to see if the error is consistent. It’s especially useful to know if the error persists with default settings or only happens under certain customizations.

If you're still having difficulties, don't hesitate to continue this thread with supplementary details based on the suggestions above. We aim to resolve your issue as best we can!

**Comment by sonali20** [comment]
Hi @glenn-jocher , thank you for your response. I was able to get it working. The image size for the prediction was set to cv::Size(640, 480). I changed it to cv::Size(640, 640) and that works. 

As for the opencv version, I had trouble getting it working with 4.6.0. Same error as https://github.com/ultralytics/ultralytics/issues/1364 and it got resolved by upgrading to 4.7.0 as others mentioned on the ticket.

Thanks for your support!

**Comment by glenn-jocher** [comment]
Hi @sonali20, glad to hear you've managed to get it working. Yes, YOLOv8 is designed to work more optimally with square inputs as the architecture assumes square inputs, so setting the image size to cv::Size(640, 640) would indeed be what would work.

Regarding OpenCV, your findings are very much appreciated. Your experience could prove to be beneficial for others who might encounter similar issues in the future. So thank you for sharing your solution, this only helps to improve and grow the YOLOv8 community and functionality.

Thanks for your support and contribution. Happy to be able to assist in this process!
